opencog/haskell/OpenCog/AtomSpace/Api.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE ForeignFunctionInterface #-}
+
+module OpenCog.AtomSpace.Api (
+      asAddNode
+    , runOnNewAtomSpace
+    , AtomSpace
+    ) where
+
+-- Note that I don't export the AtomSpace data constructors nor the asDelete/asNew functions.
+
+import Foreign
+import Foreign.C.Types
+import Foreign.C.String
+import Control.Exception
+import OpenCog.AtomSpace.Types
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Reader
+import Data.Functor
+
+-- Internal AtomSpace reference to a mutable C++ instance of the AtomSpace class.
+newtype AtomSpaceRef = AtomSpaceRef (Ptr AtomSpaceRef)
+
+-- Main Data Type for representing programs working on an AtomSpace.
+-- We have to use the IO monad because of the use of FFI for calling c functions for working
+-- on a mutable instance of the atomspace, so we have side effects.
+type AtomSpace = ReaderT AtomSpaceRef IO
+
+-- Internal functions new and delete, to create and delete C++ instances of the AtomSpace class.
+foreign import ccall "AtomSpace_CWrapper.h AtomSpace_new"
+               c_atomspace_new :: IO AtomSpaceRef
+asNew :: IO AtomSpaceRef
+asNew = c_atomspace_new
```

Comment:
Hi, Ok. I have read the code of the AtomTable class, and I think I have a general idea of how environments work. I found this wiki page: http://wiki.opencog.org/w/Multiple_atom_spaces which explains about this. I will work on this https://github.com/MarcosPividori/atomspace/issues/10!